### PR TITLE
WIP: Tables improvements

### PR DESCRIPTION
**Pull request by pdeffebach**
This PR attempts to address #24 by using `Tables.jl` versions of a few functions used in the package, notably `size`, `getindex` and iteration. 

It is mostly working, but is still a WIP. I am working on `filter`ing now. 

I will request a review from the Tables.jl authors at some point to make sure things are being done correctly.

### CODE REVIEW

**Comment by pdeffebach** [comment]
Apologies for augmenting the Project.toml. I'm not sure why that happened.

**Comment by ronisbr** [comment]
Hi @pdeffebach 

Thanks for the PR!

However, I think that your solution is not the best one. The low-level functions to print tables are designed to work with `Array`. This is native to Julia and very easy to track possible breakages. On the other hand, the Tables specification can be changed, introducing problems to every supported type on this package.

In my opinion, the correct way to do this (which is much easier) is to convert the object that is complaint with Tables.jl API to a simple Julia `Array` and call the low-level printing function.

**Comment by pdeffebach** [comment]
I understand your point, but your solution will involve allocating an array every time someone wants to print. This solution is non-allocating and might be more robust to people printing lots of times in a script. 

presumably your preferred solution is something like

```
pretty_table(table, args...; kwargs...)
    t = Tables.columns(table)
    header = collect(peopertynames(t))
    data = reduce(hcat, (getproperty(t, p) for p in header)
    pretty_print(data, args...; header = header, kwargs...)
```

**Comment by ronisbr** [comment]
In fact, it depends. For `DataFrames`, `Tables.columns` returns the same structure that is accessible using `[i,j]`. I prefer to keep the low-level functions as is and use the `Tables` as a fallback only if the type is not natively supported by PrettyTables.

There is a lot to improve about the performance of PrettyTables (and help is needed!), but I think it is better to focus on development of missing features, like LaTeX backend, before huge changes in current features. During this period, it is best to keep the low-level functions as stable as possible. After, when we have all the features we like,  then we can start tweaking for performance. Sounds like a plan?

**Comment by pdeffebach** [comment]
Do you think it would be crazy to check if a table `hasmethod` for get index by row and column? And if make a very thin wrapper that allows for `getindex` via `getproperty` from `Tables.columntable`?

**Comment by ronisbr** [comment]
In fact, this sounds awesome! Can you provide a proof of concept?

**Comment by pdeffebach** [comment]
I tihnk I will re-start this PR with a dependency on LazyArrays. This way we can get the indexing behavior you want (as though it is a matrix) without allocations. Then all code can be written generically as though we are using a matrix. 

```
 data = ApplyArray(hcat, Tables.eachcolumn(t)...)
```

**Comment by pdeffebach** [comment]
I plan on copying the code from Tables.jl [here](https://github.com/JuliaData/Tables.jl/blob/de50b30f7e393e41d56d427511c2a5ca2831d22b/src/matrix.jl#L61) for use with a Lazy Array. 

Then we can pass anyting `<:AbstractArray` around.

**Comment by ronisbr** [comment]
Please, do not plan a solution based on `LazyArrays`. As you can see in #18 we have a request to reduce the package loading time, so that it can be used as dependency to others packages. Take a look at the loading time of `LazyArrays` and `PrettyTables`:

```julia
julia> @time using PrettyTables
  0.190439 seconds (266.63 k allocations: 15.817 MiB)

julia> @time using LazyArrays
  1.697844 seconds (3.52 M allocations: 304.514 MiB)
```

Thus, it will make PrettyTables to be loaded almost 10x slower. I think we should just not worry about copying data on those specific cases. If anyone has a special use case in which the copy is not acceptable, then we can work with it.

**Comment by pdeffebach** [comment]
You are right, this is *a lot* easier. It only requires changing one line to make it all work I think. 

I will add some tests for various table types if this is good.

**Comment by ronisbr** [comment]
Wow! That's function is handy :) I liked that.

**Comment by ronisbr** [comment]
Can you please add a test at the end of file `./test/text_backend.jl` related to Issue #24?

**Comment by pdeffebach** [comment]
One final issue for writing the tests. A valid Tables interface object is a `Vector{<:NamedTuple}`. With the current interface, anything `AbstractVector` get's passed to the default printing method (no Tables stuff). 

For example

```
julia> rtable = [(x = 1, y = "a"), (x = 2, y = "b"), (x = 3, y = "c"), (x = 4, y = "d")]
julia> pretty_table(rtable)
┌──────────────────┐
│           Col. 1 │
├──────────────────┤
│ (x = 1, y = "a") │
│ (x = 2, y = "b") │
│ (x = 3, y = "c") │
│ (x = 4, y = "d") │
└──────────────────┘
```

Whereas we might want to know the user wants to work with a `Table`, in which case the result would be 

```
julia> pretty_table(Tables.columns(rtable))
┌───────┬────────┐
│     x │      y │
│ Int64 │ String │
├───────┼────────┤
│     1 │      a │
│     2 │      b │
│     3 │      c │
│     4 │      d │
└───────┴────────┘
```

We could 

1. Add a method for `pretty_table(IO::IO, data<: Vector{<:NamedTuple}` that calls `Tables.columns`
2. Let users know they have to call `Tables.columns(rtable)` to print a vector of each element form the named tuples as it's own column.

**Comment by ronisbr** [comment]
I think we can go with the second option, because, for me:

```julia
[(x = 1, y = "a"), (x = 2, y = "b"), (x = 3, y = "c"), (x = 4, y = "d")]
```

is a Vector of `NamedTuples` and the default behavior is the right one.

**Comment by ronisbr** [comment]
Hi @pdeffebach 

Two things:

1. Please, do not include the `Manifest.toml` file in the commits. This is not needed in packages.
2. It can save you a lot of time if you run `julia runtests.jl` inside the `test` folder before submitting the commits.

**Comment by pdeffebach** [comment]
Sorry about that. I do run tests locally but must have missed that. 

I added a .gitignore file so the manifest isn't included when I run `julia --project`.

File: src/print.jl [hunk 1]
```diff
@@ -352,10 +352,10 @@ function pretty_table(io::IO, table; kwargs...)
     # Get the data.
     #
     # If `table` is not compatible with Tables.jl, then an error will be thrown.
-    data = Tables.columns(table)
-
+    Tables.columns(table)
```
Comment (ronisbr) [inline src/print.jl]: Any particular reason why we need this?

Comment (ronisbr) [reply src/print.jl]: I mean, why call `Tables.columns` if we are converting to matrix below.

Comment (pdeffebach) [reply src/print.jl]: we don't need it I removed it. this PR is ready to go

**Comment by ronisbr** [comment]
@pdeffenach, please, **do not** commit `.gitignore` files.

**Comment by ronisbr** [comment]
Now it seems fine :) If the build process is successful, I will merge it. Thanks for the help!

**Comment by pdeffebach** [comment]
removed manifest.toml. Thanks for reviewing this PR! its good to be merged.

**Comment by ronisbr** [comment]
@pdeffebach The test is failing in 32-bit. I know exactly why. In your test, you are declaring a vector `[1,2,3,4]`. In 64-bit systems, it is `Int64`. However, in 32-bit, it is `Int32`. This leads to the error in 32-bit. But don't worry, I can fix after this is merged. It is very simple.

**Pull request merged**